### ISSUE TITLE
Transport boats should spawn outside the sub pen

### ISSUE THREAD

**Issue by pchote** [OpenRA/OpenRA#14065]
Having them drive through the far-too-small doors looks unpolished.

#13208 fixes this, but needs someone to take charge of rebasing and reviewing it. This is a nice polish fix, so adding to Next + 1 in the hopes that we can get it done sooner rather than later.

**Comment by GSonderling** [comment]
Can I ask, why not just increase `SpawnOffset` in `structures.yaml`?

**Comment by pchote** [comment]
Submarines should still use the doors as normal.

**Comment by GSonderling** [comment]
That does make sense. I have taken a look at this. Seems like there are only two or so merge conflicts. Few lines of code at most.

I merged the branches, resolved the conflicts and started some preliminary testing. So far so good, not a single crash, units spawn as they should and transports no longer use doors.

As far as I can tell AI acts like it did before.

**Comment by GSonderling** [comment]
It looks like there might be an issue after all. Could someone please clarify this method:

https://github.com/atlimit8/OpenRA/blob/d371aa3465b4f71507849df137a2cf05b0b57661/OpenRA.Mods.Common/Traits/Player/ClassicProductionQueue.cs#L86-L116

		protected override bool BuildUnit(ActorInfo unit)
		{
			// Find a production structure to build this actor
			var bi = unit.TraitInfo<BuildableInfo>();


			// Some units may request a specific production type, which is ignored if the AllTech cheat is enabled
			var type = developerMode.AllTech ? Info.Type : bi.BuildAtProductionType ?? Info.Type;


			var producers = self.World.ActorsWithTrait<Production>()
				.Where(x => x.Actor.Owner == self.Owner
					&& x.Trait.Info.Produces.Contains(type))
					.OrderByDescending(x => x.Actor.IsPrimaryBuilding())
					.ThenByDescending(x => x.Actor.ActorID);


			if (!producers.Any())
			{
				CancelProduction(unit.Name, 1);
				return true;
			}


			foreach (var p in producers.Where(p => !p.Actor.IsDisabled()))
			{
				if (p.Trait.Produce(p.Actor, unit, p.Trait.Faction, Info.Type))
				{
					FinishProduction();
					return true;
				}
			}


			return false;
		}


First of all the:

			var type = developerMode.AllTech ? Info.Type : bi.BuildAtProductionType ?? Info.Type;


always returns Info.Type, even if bi.BuildAtProductionType is not null. Is this supposed to happen? Because if so the null-coalescing operator (??) is pretty much meaningless. Or maybe it has some sort of purpose I don't understand. If this is a bug, it can be solved by adding brackets. Like this:

`var type = developerMode.AllTech ? Info.Type : (bi.BuildAtProductionType ?? Info.Type);`


Second question I have is about: 

 `if (p.Trait.Produce(p.Actor, unit, p.Trait.Faction, Info.Type))`

When all units used the same exit type it didn't matter that the last argument was:
`Info.Type` be default. But now it means that every unit produced at Sub Pen (and probably other buildings in mods) gets same exit type. Leading to subs not using doors etc.

Is this indented behavior or did I just mess up the merge?


If neither then I would propose something like this:
```


protected override bool BuildUnit(ActorInfo unit)
		{
			// Find a production structure to build this actor
			var bi = unit.TraitInfo<BuildableInfo>();

			// Some units may request a specific production type, which is ignored if the AllTech cheat is enabled
			var type = developerMode.AllTech ? Info.Type : (bi.BuildAtProductionType ?? Info.Type);

			var producers = self.World.ActorsWithTrait<Production>()
				.Where(x => x.Actor.Owner == self.Owner
					&& x.Trait.Info.Produces.Contains(type))
					.OrderByDescending(x => x.Actor.IsPrimaryBuilding())
					.ThenByDescending(x => x.Actor.ActorID);

			if (!producers.Any())
			{
				CancelProduction(unit.Name, 1);
				return true;
			}

			foreach (var p in producers.Where(p => !p.Actor.IsDisabled()))
			{
				if (p.Trait.Produce(p.Actor, unit, p.Trait.Faction, type))
				{
					FinishProduction();
					return true;
				}
			}

			return false;
		}

```

It with this change the units once again use correct exit type assigned to them.

**Comment by abcdefg30** [comment]
Are you sure about the `var type = developerMode.AllTech ? Info.Type : (bi.BuildAtProductionType ?? Info.Type);` change? I don't know how it could be that different without brackets. (It shouldn't always return `Info.Type`.) Anyway, they won't hurt either. Would you mind opening a PR if you managed to fix the issues of @atlimit8's one?

**Comment by GSonderling** [comment]
Honestly, I didn't think it would work either, but for some reason it does.
It can be some .NET order-of-operations quirk or maybe (although it is improbable) I mistook one variable for other during debugging. One way or another it shouldn't hurt.

As for the general functionality: Tests in RA mod showed no problems, with exceptions of those already tracked. 

However, there is an issue with cnc mod. `protected virtual ExitInfo SelectExit(Actor self, ActorInfo producee, string type, Func<ExitInfo, bool> p)` Doesn't seem to work properly for airfields. 

I do expect to encounter similar issues with other mods. Therefore I would like to wait before submitting pull request. I don't want to cause even more bugs with hot needle solution.

**Comment by GSonderling** [comment]
Turns out there was another issue hiding behind this one.

For some reason units can't find a way from airfields because `CanEnterCell()` returns false. 
Right now I'm not sure why it happens with airfields and not with normal factories. But I assume that it has something to do with unit spawning on top of the building while the building has only one defined exit. 

To make matters worse it doesn't happen every time with every unit, most often it involves bike but I have seen it happen with light tank once too. 


I found a way around this by changing https://github.com/GSonderling/OpenRA/blob/b6abaf9e0f3e2747788432b22f70d0f2dc2f5345/OpenRA.Mods.Common/Traits/Buildings/Exit.cs#L68 to `return shuffled.First();`.

This effectively gets rid of checks on accessibility of different exits. 
The crashes have stopped and the gameplay is unaffected, as far as I can tell from my experience and from how AI behaves (tested on TS, CNC, RA and D2K).


But it's also obviously sloppy solution. If anyone knows better way to solve this, I would love to hear it.
